### DISCUSSION TITLE
RTH ended sadly: DISARM not working

### DISCUSSION THREAD

**Discussion by mrechte** [iNavFlight/inav#8093] [Q&A] [answered]
Hello. It is really frustrating, having spent so much time with RTH tuning on a working drone to end up with burnt motors and controller. When hitting ground, some grass blocked the propellers and the drone started to behave erratically. The problem is that we have not been able to force disarm. See attached screen capture were, despite disarming, motors continue to spin. Parameters that were changed with CLI before flight: nav_mc_vel_z_p=70, switch_disarm_delay=0, nav_disarm_on_landing=ON. Thanks.
![Capture d’écran du 2022-06-06 18-14-37](https://user-images.githubusercontent.com/5902317/172202880-44b6909e-cbfd-49e4-86d6-a8725f95c9fd.png)

**Comment by breadoven** [comment]
How is `disarm_kill_switch` set ?

https://github.com/iNavFlight/inav/blob/master/docs/Settings.md#disarm_kill_switch

**Reply by mrechte** [reply]
It is set to the default:
```
# get disarm_kill_switch
disarm_kill_switch = ON
Allowed values: OFF, ON
```

**Comment by Jetrell** [comment]
@mrechte You may want to include your `Diff All` settings.  
Will the `Arming mode switch` disarm the flight controller in other flight modes? (e.g. Acro, Angle). Or were you able to switch out of RTH, back to another flight mode, then disarm?

The Landing detector can not tell the difference between Flight, and the propellers being tangled in the grass. The effect on the sensors is the same.
For it to work correctly, its important that the multicopter lands level, and with no obstructions to the propellers.
If you have `throttle_idle` set on the higher side. it too will increase the time taken to disarm. Especially if a wind is blowing across the propellers, increasing lift.
You should be aware, that the landing detector needs some time, before it will disarm... No one wants it to falsely disarm while flying.

**Reply by mrechte** [reply]
I am not complaining about the fact that the RTH did not complete as expected, but regarding the fact that the controller did not honour the ARM OFF mode. As you can see on the above screen capture, the flight mode is changed to ARM OFF, and the motors don't stop. In any case disarming should immediately stop the motors: it is a safety measure. Should had it worked as planned, it would have saved the drone. Honour disarm, whatever current flight mode was (actually it was FAILSAFE ON), is, IMHO, of utter most importance. I can upload the blackbox log if needed. Thanks a lot.
[INAV_cli_diff_all_20220607_183302.txt](https://github.com/iNavFlight/inav/files/8855075/INAV_cli_diff_all_20220607_183302.txt)

**Reply by Jetrell** [reply]
@mrechte In the mean time. Until a fix is implemented. You may want to avoid using `Failsafe Mode`. And only use `RTH mode`.
This doesn't mean that your quad won't respond to a signal loss failsafe. It just means that iNav will use failsafe data from the receiver to initiate RTH, and not a failsafe channel position.

As @breadoven mentioned, iNav is geared around the use of `RTH mode` and not `FS mode`. Because features can not be controller if there is no RC link.
e.g. If you had been in a true failsafe signal loss condition, and your motors got stalled in the grass. You would have no control link to disarm it with anyway.

**Comment by breadoven** [comment]
Ah, the problem is you can't disarm when in Failsafe, either caused by actual signal loss or Failsafe mode. This is to avoid false disarms caused by unreliable receiver output. Have to admit though, I'm not sure why Failsafe mode would be included in that case because the receiver signal should be reliable given it's only simulated Failsafe not something actually caused by signal problems. Does it make sense to include Failsafe mode here @DzikuVx ?

**Reply by mrechte** [reply]
A few seconds later, seeing that DISAMR was not effective, we also released the failsafe mode, and it did not help either. See attached screen capture.
![Capture d’écran du 2022-06-07 19-44-04](https://user-images.githubusercontent.com/5902317/172448432-ff49c734-b300-47fc-bad9-4218fc9b5397.png)

**Comment by breadoven** [comment]
It should have disarmed when Failsafe was turned off. I just tested it and it did exactly that. Disarm with Failsafe mode on, no disarm, switch off Failsafe and it disarmed when the sticks are moved to cancel Failsafe. I assume you moved the sticks after switching Failsafe off ?

**Reply by mrechte** [reply]
I carefully watched the log again, displaying the flightModeFlags. 
1. Before landing flightModeFlags = ARM | FAILSAFE
2. When we saw the quad struggling in the grass, switched off the ARM switch: flightModeFlags = FAILSAFE (motors continue to spin)
3. Switched off the FAILSAFE switch: flightModeFlags "0" (motors continue to spin)
4. Joystick show movements in the BB explorer (motors continue to spin)

Something is really abnormal. navState RTH_LANDING has nothing to do with that ?
Could you please direct me in the source code where this disarming logic is handled ?

**Comment by breadoven** [comment]
Relevant code is here:
https://github.com/iNavFlight/inav/blob/85edaae6463c7ebb176329ffe86281bc66cb8763/src/main/fc/rc_controls.c#L230

`flightModeFlags` just shows the switch mode status which doesn't always correspond with the mode status itself. `failsafePhase` tells you what Failsafe is actually doing, whether it's active or not.

I didn't actually test with RTH active since it shouldn't change the behaviour of the disarm switch. Disarm should work regardless of the Nav mode selected.

**Reply by mrechte** [reply]
Thanks. failsafePhase got stuck to FAILSAFE_RETURN_TO_HOME from the time FAILSAFE switch was first activated.

**Comment by breadoven** [comment]
Was that because the sticks weren't moved enough to cancel failsafe ?  rcData[0] and rcData[1] shows raw stick movement from the receiver.

**Reply by mrechte** [reply]
It seems that after above step 3, no sticks were moved (rcData[0] = rcData[1] = rcData[2] = 1504us  rcData[3] = 995us)

**Comment by mrechte** [comment]
Wouldn't it make sense to replace:
https://github.com/iNavFlight/inav/blob/85edaae6463c7ebb176329ffe86281bc66cb8763/src/main/fc/rc_controls.c#L226
with
```
if (ARMING_FLAG(ARMED) && rxIsReceivingSignal()) {
```
?
For safety reason, I would consider the DISARM switch as an emergency measure to kill the drone.
What is the point of ignoring DISARM if rxIsReceivingSignal() is true ?

**Reply by breadoven** [reply]
The logic doesn't really make sense mainly because there are 2 parts to a good link, receiving a signal and the signal being valid (channel values within expected limits). `rxIsReceivingSignal `only flags receiving a signal not if the signal is valid. Failsafe is triggered when there is no signal or it isn't valid so `rxIsReceivingSignal `could be OK (signal received) but the signal could be invalid causing a Failsafe.

The reason `rxIsReceivingSignal ` is included in the above logic is to flag a bad link before Failsafe actually triggers (there is a delay between link loss and failsafe triggering) so prevents false disarms caused by the arming channel changing value due to a poor link. However, that's the bit that doesn't make sense because it doesn't cover the situation where `rxIsReceivingSignal `is OK but the signal isn't valid potentially causing the arming channel to change value causing a false disarm. `rxIsReceivingSignal `should be accompanied with a flag for valid signal also if it's to work as required. I'm guessing `failsafeIsActive `is included because it does account for an invalid signal but it doesn't really work because it's delayed for some time after the signal is flagged as invalid.

**Reply by rechtem** [reply]
In that case, it is just a matter of adding the test to make sure there signal is valid. Also there is another safety measure with the nav_mc|fw_auto_disarm_delay parameter, if not set to 0, will make sure the signal is stable for the set delay ?

**Comment by breadoven** [comment]
I'd forgotten that sensibly you still need `failsafeIsActive()` as a basic protection against false disarm mainly to avoid problems when the signal is intermittent. Failsafe locks out the intermittent behaviour allowing the pilot to wait until the signal has become more reliable before cancelling the failsafe using the sticks. I suppose it's a trade off between avoiding false disarms and intentional disarming being a bit slower in certain circumstances.

You shouldn't really use Failsafe mode for RTH just use RTH mode itself. If you do use Failsafe mode then switch it off once it's triggered and it will remain active only requiring the sticks to be moved to cancel it. I only use Failsafe mode to test failsafe behaviour without the risks involved in actually turning the transmitter off.

**Reply by mrechte** [reply]
OK I see, you mean use RTH mode instead of FAILSAFE to test the procedure. So FAILSAFE mode is a nuisance: if one wants to test the failsafe, it is just a matter of  putting off the radio. This behaviour should be clearly explained in the doc. I may propose a PR in that direction if the project wishes so. Thanks a lot.

**Comment by breadoven** [comment, marked as answer]
Some people seem to use Failsafe mode set to RTH to trigger a RTH when in fact it's better to just use RTH mode itself if all you want to do is RTH.

Failsafe mode can be used to test the set failsafe procedure is triggered correctly. It's safer than just switching off the transmitter because the channels are still received normally avoiding any surprises caused by changes in channel values which can occur when the signal is lost (failsafe channel behaviour can be checked in Configurator Receiver tab). The ultimate failsafe test is to switch off the transmitter but I'd only do this on the ground to check for unintended disarm.

The only other reason for using Failsafe mode is as a means of triggering failsafe if the signal is actually lost as explained in Option 2 of the following Wiki:

https://github.com/iNavFlight/inav/wiki/Failsafe

But this shouldn't be necessary on most newer RC equipment.
